src/action/common/place_nix_configuration.rs

```diff
@@ -28,65 +28,191 @@ impl PlaceNixConfiguration {
+        nix_conf_insert_settings.push("!include ./defaults.conf".into());
+        nix_conf_insert_settings.extend(extra_conf);
+        let nix_conf_insert_fragment = nix_conf_insert_settings.join("\n");
 
-        settings.insert(
-            "bash-prompt-prefix".to_string(),
-            "(nix:$name)\\040".to_string(),
-        );
+        let mut defaults_conf_settings = vec![
+            ("build-users-group", nix_build_group_name),
+            (
+                "experimental-features",
+                "nix-command flakes auto-allocate-uids".into(),
+            ),
+        ];
+        // https://github.com/DeterminateSystems/nix-installer/issues/449#issuecomment-1551782281
+        defaults_conf_settings.push(("bash-prompt-prefix", "(nix:$name)\\040".into()));
+        defaults_conf_settings.push(("extra-nix-path", "nixpkgs=flake:nixpkgs".into()));
         if let Some(ssl_cert_file) = ssl_cert_file {
             let ssl_cert_file_canonical = ssl_cert_file
                 .canonicalize()
                 .map_err(|e| Self::error(ActionErrorKind::Canonicalize(ssl_cert_file, e)))?;
-            settings.insert(
-                "ssl-cert-file".to_string(),
+            defaults_conf_settings.push((
+                "ssl-cert-file = {}",
                 ssl_cert_file_canonical.display().to_string(),
-            );
+            ));
         }
-        settings.insert(
-            "extra-nix-path".to_string(),
-            "nixpkgs=flake:nixpkgs".to_string(),
-        );
-
+        #[cfg(not(target_os = "macos"))]
+        defaults_conf_settings.push(("auto-optimise-store", "true".into()));
         // Auto-allocate uids is broken on Mac. Tools like `whoami` don't work.
         // e.g. https://github.com/NixOS/nix/issues/8444
         #[cfg(not(target_os = "macos"))]
-        settings.insert("auto-allocate-uids".to_string(), "true".to_string());
+        defaults_conf_settings.push(("auto-allocate-uids", "true".into()));
+        let defaults_conf_insert_fragment = defaults_conf_settings
+            .iter()
+            .map(|(s, v)| format!("{s} = {v}"))
+            .collect::<Vec<_>>()
+            .join("\n");
+
+        let nix_conf_insert_position = if Path::new(NIX_CONF).exists() {
+            let existing_nix_conf = tokio::fs::read_to_string(NIX_CONF)
+                .await
+                .map_err(|e| ActionErrorKind::Read(NIX_CONF.into(), e))
+                .map_err(Self::error)?;
+            tracing::trace!("Found existing `/etc/nix/nix.conf`");
+
+            // Find the first line that isn't just `# ...` comments
+            let mut chosen_insert_after = None;
+            // Warn if there seems to be a setting which is a duplicate of one we set.
+            let mut existing_conf_settings = vec![];
+
+            for (index, line) in existing_nix_conf.lines().enumerate() {
+                let line = line.trim();
+                if line.is_empty() || line.starts_with('#') {
```

Comment:
nit:

if we're looking for the first line that isn't a comment, then wouldn't an empty line suffice?

Reply:
Yes and we check it with `line.is_empty()`, however it's possible someone does

```ini
# Stuff here
auto-allocate-uids = true
```

In which case we want it after `# Stuff here` and before anything else.